### PR TITLE
Support semi-sync without async fallback.

### PR DESCRIPTION
**Pull request by enisoc**
@alainjobart 

When using semi-sync with rpl_semi_sync_master_wait_no_slave=1 and
rpl_semi_sync_master_timeout=[effectively infinite], it's necessary to
toggle master-side semi-sync at the proper times to prevent slaves from
getting stuck (since they have no slaves of their own to ACK).

This commit adds a new vttablet flag -enable_semi_sync, which causes the
replication-management features of Vitess (e.g. InitShardMaster and
various reparent operations) to automatically manage semi-sync settings.

With these settings, it becomes (reasonably) safe to reparent away from
an unresponsive master, as long as the following can be verified:

* All semi-sync slaves have stopped replicating from the old master.
* The chosen new master is the farthest ahead in replication.

Currently we only allow slaves marked in Vitess terms as "replica"
tablets to be semi-sync slaves (i.e. to send ACKs). Vitess slaves marked as
"rdonly" will NOT ACK, because they are not eligible to become masters,
so we never want them to be the only slave with the latest update.
This means when using semi-sync, every shard must have at least 2 live
"replica" type tablets (including the master) in order to accept writes.

[Review on Reviewable](https://reviewable.io/reviews/youtube/vitess/1521)

### CODE REVIEW

**Comment by alainjobart** [comment]
LGTM thanks for the detailed work!

File: go/vt/mysqlctl/mysql_daemon.go [hunk 1]
```diff
@@ -438,3 +440,13 @@ func (fmd *FakeMysqlDaemon) GetAppConnection() (dbconnpool.PoolConnection, error
...
 }
+
+// SetSemiSyncEnabled is part of the MysqlDaemon interface.
+func (fmd *FakeMysqlDaemon) SetSemiSyncEnabled(master, slave bool) error {
+	return nil
+}
+
+// SemiSyncEnabled is part of the MysqlDaemon interface.
+func (fmd *FakeMysqlDaemon) SemiSyncEnabled() (master, slave bool) {
+	return false, false
```
Comment (alainjobart) [inline go/vt/mysqlctl/mysql_daemon.go]: Could we remember the values, return the right ones, and test them in the wrangler/testlib tests?

Comment (enisoc) [inline go/vt/mysqlctl/mysql_daemon.go]: Done.

**Comment by alainjobart** [comment]
LGTM

**Comment by michael-berlin** [comment]
Three small comments from my side.

---

Reviewed 3 of 7 files at r1, 1 of 1 files at r2, 14 of 14 files at r3, 2 of 2 files at r4, 9 of 9 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[go/vt/mysqlctl/replication.go, line 381 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1521#-KBQHuZMpx0XIJ_kwy9J:-KBQHuZMpx0XIJ_kwy9K:-1657154626) ([raw file](https://github.com/youtube/vitess/blob/43b007dbb826319c812c9b81e8d93c786bd78933/go/vt/mysqlctl/replication.go#L381)):*
No need to check of the entry exists in the map.

A non existent entry will return a zero value :)

---

*[go/vt/tabletmanager/rpc_replication.go, line 26 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1521#-KBQO3SM3rcTJALJvlLR:-KBQO3SM3rcTJALJvlLS:1455272930) ([raw file](https://github.com/youtube/vitess/blob/43b007dbb826319c812c9b81e8d93c786bd78933/go/vt/tabletmanager/rpc_replication.go#L26)):*
FYI: Instead of exporting it, you could also do this in the test:

```go
flag.Set("enable_semi_sync", "true")
```

---

*[test/mysql_flavor.py, line 38 \[r5\]](https://reviewable.io:443/reviews/youtube/vitess/1521#-KBVDNmupR4yvrkiyQK7:-KBVDNmupR4yvrkiyQK8:-458749010) ([raw file](https://github.com/youtube/vitess/blob/43b007dbb826319c812c9b81e8d93c786bd78933/test/mysql_flavor.py#L38)):*
slave instead of master here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/youtube/vitess/1521#-:-KBVhdMCB_d7WTwDIy7E:-1409204435)*

**Pull request merged**